pkg/provider/ngrok/provider.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright © 2025 ESO Maintainer Team
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ngrok
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	esv1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1"
+	"github.com/external-secrets/external-secrets/pkg/utils/resolvers"
+	"github.com/ngrok/ngrok-api-go/v7"
+	"github.com/ngrok/ngrok-api-go/v7/secrets"
+	"github.com/ngrok/ngrok-api-go/v7/vaults"
+)
+
+var (
+	defaultAPIURL = "https://api.ngrok.com"
+	userAgent     = "external-secrets"
+
+	errClusterStoreRequiresNamespace = errors.New("cluster store requires namespace")
+	errInvalidStore                  = errors.New("invalid store")
+	errInvalidStoreSpec              = errors.New("invalid store spec")
+	errInvalidStoreProv              = errors.New("invalid store provider")
+	errInvalidNgrokProv              = errors.New("invalid ngrok provider")
+	errInvalidAuthAPIKeyRequired     = errors.New("ngrok provider auth APIKey is required")
+	errInvalidAPIURL                 = errors.New("invalid API URL")
+	errMissingVaultName              = errors.New("ngrok provider vault name is required")
+)
+
+type vaultClientFactory func(cfg *ngrok.ClientConfig) VaultClient
+type secretsClientFactory func(cfg *ngrok.ClientConfig) SecretsClient
+
+var getVaultsClient vaultClientFactory = func(cfg *ngrok.ClientConfig) VaultClient {
+	return vaults.NewClient(cfg)
+}
+
+var getSecretsClient secretsClientFactory = func(cfg *ngrok.ClientConfig) SecretsClient {
+	return secrets.NewClient(cfg)
+}
+
+type Provider struct{}
+
+// Capabilities returns the provider supported capabilities (ReadOnly, WriteOnly, ReadWrite). Currently,
+// ngrok only supports WriteOnly capabilities.
+func (p *Provider) Capabilities() esv1.SecretStoreCapabilities {
+	return esv1.SecretStoreWriteOnly
+}
+
+// NewClient implements the Client interface.
+func (p *Provider) NewClient(ctx context.Context, store esv1.GenericStore, kubeClient kubeClient.Client, namespace string) (esv1.SecretsClient, error) {
```

Comment:
shouldnt this provider have lookup timeouts? whats the default behavior?

Reply:
Are you asking about a lookup timeout for finding the vault ID by its name? If so, it just uses the provided context when iterating over the vaults in the ngrok API.